In case of CNS_NEW_SYNC, Volume expansion to succeed when SPS is down.

Aishwarya-Hebbar opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**: Modifying volume expansion testcase to succeed when SPS is down in case of CNS_NEW_SYNC is enabled.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes bug 2865342
https://bugzilla.eng.vmware.com/show_bug.cgi?id=2865342

**Special notes for your reviewer**:

**Release note**:

1. Vanilla + WCP : Verify online volume expansion when SPS-Service is down
2. WCP: Verify Offline volume expansion when SPS-Service is down

Testing:
Vanilla: 
with cns_new_sync: https://gist.github.com/Aishwarya-Hebbar/4c53ad84165787742110d829bbe56a48
without cns_new_sync: https://gist.github.com/Aishwarya-Hebbar/988ffddfeb1dfbefc22b34eb81748ab1
WCP:
with cns_new_sync: In progress
without cns_new_sync: https://gist.github.com/Aishwarya-Hebbar/a1cd691df3814283ec5d5d92fd2eae59

svcbot-qecnsdp commented:
Started vanilla Block pipeline... Build Number: 770

svcbot-qecnsdp commented:
``` 
Block vanilla build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 
JUnit report was created: /home/worker/workspace/Block-Vanilla/Results/770/vsphere-csi-driver/tests/e2e/junit.xml

Ran 1 of 262 Specs in 187.277 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 261 Skipped
PASS

Ginkgo ran 1 suite in 4m33.215929836s
Test Suite Passed

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 231

svcbot-qecnsdp commented:
``` 
Build ID: 231
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
Ran 2 of 262 Specs in 272.292 seconds
FAIL! -- 1 Passed | 1 Failed | 0 Pending | 260 Skipped
--- FAIL: TestE2E (272.41s)
FAIL

Ginkgo ran 1 suite in 5m12.087690981s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/csi-wcp-pre-check-in/Results/231/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 232

svcbot-qecnsdp commented:
``` 
Build ID: 232
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
Ran 2 of 262 Specs in 280.511 seconds
FAIL! -- 1 Passed | 1 Failed | 0 Pending | 260 Skipped
--- FAIL: TestE2E (280.63s)
FAIL

Ginkgo ran 1 suite in 5m18.457385908s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/csi-wcp-pre-check-in/Results/232/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 234

svcbot-qecnsdp commented:
``` 
Build ID: 234
WCP build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 

Ran 2 of 262 Specs in 275.900 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 260 Skipped
PASS

Ginkgo ran 1 suite in 5m12.326092595s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/csi-wcp-pre-check-in/Results/234/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 235

svcbot-qecnsdp commented:
Started Vanilla block pre-checkin pipeline... Build Number: 537

svcbot-qecnsdp commented:
``` 
Build ID: 235
WCP build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 

Ran 14 of 262 Specs in 3133.228 seconds
SUCCESS! -- 14 Passed | 0 Failed | 0 Pending | 248 Skipped
PASS

Ginkgo ran 1 suite in 52m49.840781728s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/csi-wcp-pre-check-in/Results/235/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
``` 
Build ID: 537
Block vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
JUnit report was created: /home/worker/workspace/csi-block-vanilla-pre-check-in/537/vsphere-csi-driver/tests/e2e/junit.xml

Ran 1 of 262 Specs in 322.307 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 261 Skipped
PASS

Ginkgo ran 1 suite in 6m14.418284026s
Test Suite Passed
--
JUnit report was created: /home/worker/workspace/csi-block-vanilla-pre-check-in/537/vsphere-csi-driver/tests/e2e/junit.xml

Ran 12 of 262 Specs in 6485.381 seconds
SUCCESS! -- 12 Passed | 0 Failed | 0 Pending | 250 Skipped
PASS

Ginkgo ran 1 suite in 1h48m19.833241899s
Test Suite Passed
--
/home/worker/workspace/csi-block-vanilla-pre-check-in/537/vsphere-csi-driver/tests/e2e/svmotion_detached_volume.go:164

Ran 39 of 262 Specs in 1188.075 seconds
FAIL! -- 37 Passed | 2 Failed | 0 Pending | 223 Skipped


Ginkgo ran 1 suite in 20m2.312875274s
Test Suite Failed

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 4

svcbot-qecnsdp commented:
``` 
Build ID: 4
WCP build status: FAILURE 
Stage before exit: null 

```

rpanduranga reviewed (approved):
Do not share internal sites details on the github comments as the link to open source community will not be accessible.  Instead provide all the required details in the format of text .

Use https://gist.github.com/ to store your logs and provide the gist url in the testing done section 
Jenkins URL will be mostly lost after week, so you should keep this evidences in the permanent storage locations.

//Approving ahead requesting to modify the github descriptions

kavyashree-r commented on tests/e2e/vsphere_volume_expansion.go:
```diff
@@ -717,21 +716,6 @@ var _ = ginkgo.Describe("Volume Expansion Test", func() {
 		gomega.Expect(err).NotTo(gomega.HaveOccurred())
 		gomega.Expect(pvclaim).NotTo(gomega.BeNil())
 
-		ginkgo.By("File system resize should not succeed Since SPS service is down. Expect an error")
```
If you delete these lines 720 - 734 , The regular tests which runs in BR pipeline will fail,  
Make a check , If it is Supervisor and if CNS_NEW_SYNC feature is enabled , then only expect the Volume expansion to complete without SPS service

kavyashree-r commented on tests/e2e/vsphere_volume_expansion.go:
```diff
@@ -1463,17 +1452,6 @@ var _ = ginkgo.Describe("Volume Expansion Test", func() {
 		gomega.Expect(err).NotTo(gomega.HaveOccurred())
 		gomega.Expect(pvclaim).NotTo(gomega.BeNil())
 
-		ginkgo.By("File system resize should not succeed Since SPS service is down. Expect an error")
```
Same as above comment . Please add a check for CNS_NEW_SYnc feature enabled or not , then take the decision, whether the volume expansion should complete or not

svcbot-qecnsdp commented:
Started vanilla Block pipeline... Build Number: 2

svcbot-qecnsdp commented:
``` 
Block vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
# sigs.k8s.io/vsphere-csi-driver/v2/tests/e2e [sigs.k8s.io/vsphere-csi-driver/v2/tests/e2e.test]
tests/e2e/vsphere_volume_expansion.go:735:5: undefined: expectedErrMsg
tests/e2e/vsphere_volume_expansion.go:737:5: undefined: expectedErrMsg
tests/e2e/vsphere_volume_expansion.go:739:52: undefined: expectedErrMsg
tests/e2e/vsphere_volume_expansion.go:740:47: undefined: expectedErrMsg

Ginkgo ran 1 suite in 1m2.896302464s
Test Suite Failed

```

svcbot-qecnsdp commented:
Started vanilla Block pipeline... Build Number: 3

svcbot-qecnsdp commented:
``` 
Block vanilla build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 
JUnit report was created: /home/worker/workspace/block-vanilla-kai-fss/Results/3/vsphere-csi-driver/tests/e2e/junit.xml

Ran 1 of 268 Specs in 181.886 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 267 Skipped
PASS

Ginkgo ran 1 suite in 4m15.00685271s
Test Suite Passed

```

svcbot-qecnsdp commented:
Started vanilla Block pipeline... Build Number: 4

svcbot-qecnsdp commented:
``` 
Block vanilla build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 
JUnit report was created: /home/worker/workspace/block-vanilla-kai-fss/Results/4/vsphere-csi-driver/tests/e2e/junit.xml

Ran 1 of 268 Specs in 419.318 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 267 Skipped
PASS

Ginkgo ran 1 suite in 8m14.630480331s
Test Suite Passed

```

Aishwarya-Hebbar commented:
Addressed review comments

> Do not share internal sites details on the github comments as the link to open source community will not be accessible. Instead provide all the required details in the format of text .
> 
> Use https://gist.github.com/ to store your logs and provide the gist url in the testing done section Jenkins URL will be mostly lost after week, so you should keep this evidences in the permanent storage locations.
> 
> //Approving ahead requesting to modify the github descriptions

Changed github description

kavyashree-r commented on tests/e2e/vsphere_volume_expansion.go:
```diff
@@ -651,6 +651,8 @@ var _ = ginkgo.Describe("Volume Expansion Test", func() {
 		var err error
 		var expectedErrMsg string
+		vcAddress := e2eVSphere.Config.Global.VCenterHostname + ":" + sshdPort
+		featureEnabled := isFssEnabled(vcAddress, "CNS_NEW_SYNC")
```
It is better if we read CNS_NEW_SYNC as a constant. please add it in e2e_common.go

kavyashree-r commented on tests/e2e/vsphere_volume_expansion.go:
```diff
@@ -651,6 +651,8 @@ var _ = ginkgo.Describe("Volume Expansion Test", func() {
+		featureEnabled := isFssEnabled(vcAddress, "CNS_NEW_SYNC")
 
 		volHandle, pvclaim, pv, storageclass := createSCwithVolumeExpansionTrueAndDynamicPVC(
 			f, client, "", storagePolicyName, namespace)
```
In defer block , Line number 660  and 1446 , We should not delete storage class in case of supervisor cluster

sashrith commented on tests/e2e/util.go:
```diff
@@ -1168,6 +1168,18 @@ func invokeVCenterServiceControl(command, service, host string) error {
+	sshCmd := fmt.Sprintf("python /usr/sbin/feature-state-wrapper.py %s", fss)
+	framework.Logf("Checking if fss is enabled on vCenter host %v", host)
+	result, err := fssh.SSH(sshCmd, host, framework.TestContext.Provider)
+	if err == nil || result.Code == 0 {
```
`if err == nil && result.Code == 0`

sashrith commented on tests/e2e/util.go:
```diff
@@ -1168,6 +1168,18 @@ func invokeVCenterServiceControl(command, service, host string) error {
+	if err == nil || result.Code == 0 {
+		//fssh.LogResult(result)
+		return strings.Replace(result.Stdout, "\n", "", -1) == "enabled"
+	}
```
add an else block and propagate the error

sashrith commented on tests/e2e/util.go:
```diff
@@ -1168,6 +1168,18 @@ func invokeVCenterServiceControl(command, service, host string) error {
+	framework.Logf("Checking if fss is enabled on vCenter host %v", host)
+	result, err := fssh.SSH(sshCmd, host, framework.TestContext.Provider)
+	if err == nil || result.Code == 0 {
+		//fssh.LogResult(result)
```
you can keep this

sashrith commented on tests/e2e/vsphere_volume_expansion.go:
```diff
@@ -717,20 +718,22 @@ var _ = ginkgo.Describe("Volume Expansion Test", func() {
-		err = waitForEvent(ctx, client, namespace, expectedErrMsg, pvclaim.Name)
-		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		if !featureEnabled {
+			ginkgo.By("File system resize should not succeed Since SPS service is down. Expect an error")
```
Since -> since,
Expect -> Expecting

sashrith commented on tests/e2e/vsphere_volume_expansion.go:
```diff
@@ -1495,12 +1512,12 @@ var _ = ginkgo.Describe("Volume Expansion Test", func() {
 		if queryResult.Volumes[0].BackingObjectDetails.(*cnstypes.CnsBlockBackingDetails).CapacityInMb != newSizeInMb {
 			err = fmt.Errorf("got wrong disk size after volume expansion")
 		}
-		gomega.Expect(err).NotTo(gomega.HaveOccurred())
```
we are creating an error for wrong disk size after volume expansion above, we should keep the check `gomega.Expect(err).NotTo(gomega.HaveOccurred())`

sashrith commented on tests/e2e/util.go:
```diff
@@ -1168,6 +1168,18 @@ func invokeVCenterServiceControl(command, service, host string) error {
 	return nil
 }
 
+//checks if vCenter has a particular FSS enabled or
```
recommended format of the comment,
// <func_name> <purpose>

FSS enabled or 'not'

svcbot-qecnsdp commented:
``` 
Block vanilla build status: FAILURE 
Stage before exit: checkout 

```

svcbot-qecnsdp commented:
Started vanilla Block pipeline... Build Number: 6

svcbot-qecnsdp commented:
``` 
Block vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
hack/run-e2e-test.sh
Failed to compile e2e:

# sigs.k8s.io/vsphere-csi-driver/v2/tests/e2e [sigs.k8s.io/vsphere-csi-driver/v2/tests/e2e.test]
tests/e2e/util.go:1238:2: syntax error: unexpected else, expecting }

Ginkgo ran 1 suite in 34.528410925s
Test Suite Failed

```

svcbot-qecnsdp commented:
Started vanilla Block pipeline... Build Number: 7

svcbot-qecnsdp commented:
``` 
Block vanilla build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 
JUnit report was created: /home/worker/workspace/block-vanilla-kai-fss/Results/7/vsphere-csi-driver/tests/e2e/junit.xml

Ran 1 of 272 Specs in 220.005 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 271 Skipped
PASS

Ginkgo ran 1 suite in 4m33.750198568s
Test Suite Passed

```

sashrith commented on tests/e2e/e2e_common.go:
```diff
@@ -31,6 +31,7 @@ const (
 	adminPassword                              = "Admin!23"
 	busyBoxImageOnGcr                          = "gcr.io/google_containers/busybox:1.27"
 	nginxImage                                 = "k8s.gcr.io/nginx-slim:0.8"
+	cnsNewSync                                 = "CNS_NEW_SYNC"
```
`cnsNewSyncFSS` might be better

sashrith commented on tests/e2e/util.go:
```diff
@@ -1168,6 +1168,23 @@ func invokeVCenterServiceControl(command, service, host string) error {
+	result, err := fssh.SSH(sshCmd, host, framework.TestContext.Provider)
+	fssh.LogResult(result)
+	if err == nil && result.Code == 0 {
+		return strings.Replace(result.Stdout, "\n", "", -1) == "enabled"
```
`strings.TrimSpace()` might be more readable

sashrith commented on tests/e2e/util.go:
```diff
@@ -1168,6 +1168,23 @@ func invokeVCenterServiceControl(command, service, host string) error {
+	} else {
+		ginkgo.By(fmt.Sprintf("couldn't execute command: %s on vCenter host: %v", sshCmd, err))
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
```
remove extra line

sashrith reviewed (approved):
Few minor comments, approving ahead

kavyashree-r commented:
/approve

svcbot-qecnsdp commented:
Started Vanilla block pre-checkin pipeline... Build Number: 616

svcbot-qecnsdp commented:
``` 
Build ID: 616
Block vanilla build status: FAILURE 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
JUnit report was created: /home/worker/workspace/csi-block-vanilla-pre-check-in@3/616/vsphere-csi-driver/tests/e2e/junit.xml

Ran 1 of 274 Specs in 358.335 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 273 Skipped
PASS

Ginkgo ran 1 suite in 6m39.493549281s
Test Suite Passed
--
JUnit report was created: /home/worker/workspace/csi-block-vanilla-pre-check-in@3/616/vsphere-csi-driver/tests/e2e/junit.xml

Ran 12 of 274 Specs in 6622.754 seconds
SUCCESS! -- 12 Passed | 0 Failed | 0 Pending | 262 Skipped
PASS

Ginkgo ran 1 suite in 1h50m33.485690347s
Test Suite Passed
--
/home/worker/workspace/csi-block-vanilla-pre-check-in@3/616/vsphere-csi-driver/tests/e2e/csi_cns_telemetry.go:203

Ran 40 of 274 Specs in 693.179 seconds
FAIL! -- 39 Passed | 1 Failed | 0 Pending | 234 Skipped


Ginkgo ran 1 suite in 11m44.465199176s
Test Suite Failed

```

svcbot-qecnsdp commented:
Started Vanilla block pre-checkin pipeline... Build Number: 617

svcbot-qecnsdp commented:
``` 
Build ID: 617
Block vanilla build status: SUCCESS 
Stage before exit: e2e-tests 
Jenkins E2E Test Results: 
JUnit report was created: /home/worker/workspace/csi-block-vanilla-pre-check-in@2/617/vsphere-csi-driver/tests/e2e/junit.xml

Ran 1 of 274 Specs in 162.883 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 273 Skipped
PASS

Ginkgo ran 1 suite in 3m24.664639063s
Test Suite Passed

```

chethanv28 reviewed (approved):
/ok-to-test
/approve

chethanv28 commented:
/lgtm
